ssl/quic/quic_port.c

```diff
@@ -739,6 +747,111 @@ static int marshal_validation_token(QUIC_VALIDATION_TOKEN *token,
     return 1;
 }
 
+/**
+ * @brief Encrypts a validation token using AES-256-GCM
+ *
+ * @param port       The QUIC port containing the encryption key
+ * @param plaintext  The data to encrypt
+ * @param pt_len     Length of the plaintext
+ * @param ciphertext Buffer to receive encrypted data. If NULL, ct_len will be
+ *                   set to the required buffer size and function returns
+ *                   immediately.
+ * @param ct_len     Pointer to size_t that will receive the ciphertext length.
+ *                   This also includes bytes for QUIC_RETRY_INTEGRITY_TAG_LEN.
+ *
+ * @return 1 on success, 0 on failure
+ *
+ * The ciphertext format is:
+ * [EVP_GCM_IV_LEN bytes IV][encrypted data][EVP_GCM_TAG_LEN bytes tag]
+ */
+static int encrypt_validation_token(const QUIC_PORT *port,
+                                    const unsigned char *plaintext,
+                                    size_t pt_len,
+                                    unsigned char *ciphertext,
+                                    size_t *ct_len)
+{
+    EVP_CIPHER_CTX *ctx = NULL;
+    int len, ret = 0;
+    const EVP_CIPHER *cipher = NULL;
+    unsigned char *iv = ciphertext;
+    unsigned char *data = ciphertext + EVP_GCM_IV_LEN;
+    unsigned char *tag = ciphertext + EVP_GCM_IV_LEN + pt_len;
+
+    *ct_len = EVP_GCM_IV_LEN + pt_len + EVP_GCM_TAG_LEN
+              + QUIC_RETRY_INTEGRITY_TAG_LEN;
+    if (ciphertext == NULL)
+        return 1;
+
+    if ((ctx = EVP_CIPHER_CTX_new()) == NULL
+        || (cipher = EVP_CIPHER_fetch(port->engine->libctx,
+                                      "AES-256-GCM", NULL)) == NULL
+        || !RAND_bytes_ex(port->engine->libctx, ciphertext, EVP_GCM_IV_LEN, 0)
+        || !EVP_EncryptInit_ex(ctx, cipher, NULL, port->token_key, iv)
+        || !EVP_EncryptUpdate(ctx, data, &len, plaintext, pt_len)
+        || !EVP_EncryptFinal_ex(ctx, data + pt_len, &len)
+        || !EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_GET_TAG, EVP_GCM_TAG_LEN, tag))
+        goto err;
+
+    ret = 1;
+err:
+    EVP_CIPHER_CTX_free(ctx);
+    return ret;
+}
+
+/**
+ * @brief Decrypts a validation token using AES-256-GCM
+ *
+ * @param port       The QUIC port containing the decryption key
+ * @param ciphertext The encrypted data (including IV and tag)
+ * @param ct_len     Length of the ciphertext
+ * @param plaintext  Buffer to receive decrypted data. If NULL, pt_len will be
+ *                   set to the required buffer size.
+ * @param pt_len     Pointer to size_t that will receive the plaintext length
+ *
+ * @return 1 on success, 0 on failure
+ *
+ * Expected ciphertext format:
+ * [EVP_GCM_IV_LEN bytes IV][encrypted data][EVP_GCM_TAG_LEN bytes tag]
+ */
+static int decrypt_validation_token(const QUIC_PORT *port,
+                                  const unsigned char *ciphertext,
+                                  size_t ct_len,
+                                  unsigned char *plaintext,
+                                  size_t *pt_len)
```

Comment:
style nit: these should be moved over to line up with the arguments on the first line